library/src/main/java/org/apache/apex/malhar/lib/fs/s3/S3RecordReader.java

```diff
@@ -0,0 +1,545 @@
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.apex.malhar.lib.fs.s3;
+
+import java.io.IOException;
+import java.util.Arrays;
+
+import org.apache.apex.malhar.lib.fs.FSRecordReader;
+import org.apache.hadoop.classification.InterfaceStability.Evolving;
+import org.apache.hadoop.fs.FSDataInputStream;
+
+import com.amazonaws.auth.BasicAWSCredentials;
+import com.amazonaws.services.s3.AmazonS3;
+import com.amazonaws.services.s3.AmazonS3Client;
+import com.amazonaws.services.s3.model.GetObjectRequest;
+import com.amazonaws.services.s3.model.S3Object;
+import com.amazonaws.services.s3.model.S3ObjectInputStream;
+import com.esotericsoftware.kryo.NotNull;
+import com.google.common.base.Preconditions;
+import com.google.common.io.ByteStreams;
+
+import com.datatorrent.api.Context.OperatorContext;
+import com.datatorrent.lib.io.block.BlockMetadata;
+import com.datatorrent.lib.io.block.ReaderContext;
+
+/**
+ * This operator can be used for reading records/tuples from S3 in parallel
+ * (without ordering guarantees between tuples). Records can be delimited (e.g.
+ * newline) or fixed width records. Output tuples are byte[].
+ *
+ * Typically, this operator will be connected to output of FileSplitterInput to
+ * read records in parallel.
+ */
+@Evolving
+public class S3RecordReader extends FSRecordReader
+{
+  private transient AmazonS3 s3Client;
+  private String endPoint;
+  @NotNull
+  private String bucketName;
+  @NotNull
+  private String accessKey;
+  @NotNull
+  private String secretAccessKey;
+  private int overflowBufferSize;
+
+  public S3RecordReader()
+  {
+    /*
+     * Set default overflowBufferSize to 1MB
+     */
+    overflowBufferSize = 1024 * 1024;
+  }
+
+  /**
+   * Extract the file path from given block and set it to the readerContext
+   *
+   * @param block
+   *          block metadata
+   * @return stream
+   * @throws IOException
+   */
+  @Override
+  protected FSDataInputStream setupStream(BlockMetadata.FileBlockMetadata block) throws IOException
+  {
+    String filePath = block.getFilePath();
+    // File path would be the path after bucket name.
+    // Check if the file path starts with "/"
+    if (filePath.startsWith("/")) {
+      filePath = filePath.substring(1);
+    }
+
+    RECORD_READER_MODE recordReaderMode = RECORD_READER_MODE.valueOf(this.getMode());
+    if (recordReaderMode == RECORD_READER_MODE.FIXED_WIDTH_RECORD) {
+      ((S3FixedWidthRecordReaderContext)readerContext).setFilePath(filePath);
+      ((S3FixedWidthRecordReaderContext)readerContext).setFileLength(block.getFileLength());
+    } else if (recordReaderMode == RECORD_READER_MODE.DELIMITED_RECORD) {
+      ((S3DelimitedRecordReaderContext)readerContext).setFilePath(filePath);
+      ((S3DelimitedRecordReaderContext)readerContext).setFileLength(block.getFileLength());
+    }
+    return null;
+  }
+
+  /**
+   * Initialize the reader context
+   */
```

Comment:
The parent method also initializes the reader context, I added the additional functionality of initializing s3Client